pkg/kubelet/kubelet_pods.go

```diff
@@ -1099,46 +1117,170 @@ func (kl *Kubelet) HandlePodCleanups() error {
 	}
 
 	// Remove any orphaned mirror pods.
-	kl.podManager.DeleteOrphanedMirrorPods()
+	kl.deleteOrphanedMirrorPods()
 
 	// Remove any cgroups in the hierarchy for pods that are no longer running.
 	if kl.cgroupsPerQOS {
-		kl.cleanupOrphanedPodCgroups(cgroupPods, activePods)
+		kl.cleanupOrphanedPodCgroups(pcm, cgroupPods, activePods)
 	}
 
 	kl.backOff.GC()
 	return nil
 }
 
-// podKiller launches a goroutine to kill a pod received from the channel if
+// PodKiller handles requests for killing pods
+type PodKiller interface {
+	// KillPod receives pod speficier representing the pod to kill
+	KillPod(pair *kubecontainer.PodPair)
+	// PerformPodKillingWork performs the actual pod killing work via calling CRI
+	// It returns after currently outstanding pod killing requests are served
+	PerformPodKillingWork()
+	// IsPodPendingTerminationByPodName checks whether the pod for the given full pod name is pending termination
+	IsPodPendingTerminationByPodName(podFullname string) bool
+	// IsPodPendingTerminationByUID checks whether the pod for the given uid is pending termination
+	IsPodPendingTerminationByUID(uid types.UID) bool
+}
+
+// podKillerWithSlice is an implementation of PodKiller which receives pod killing requests via slice
+type podKillerWithSlice struct {
+	// slice for storing pods to kill.
+	podsToKill []*kubecontainer.PodPair
+	// lock for synchronization between HandlePodCleanups and pod killer
+	podKillingLock *sync.Mutex
+	// used to notify the arrival of pod killing request
+	cond *sync.Cond
+	// podTerminationMap keeps track of the progress of pod termination
+	// The key is the UID of the pod and the value is the full name of the pod
+	podTerminationMap map[string]string
+	// killPod is the func which invokes runtime to kill the pod
+	killPod func(pod *v1.Pod, runningPod *kubecontainer.Pod, status *kubecontainer.PodStatus, gracePeriodOverride *int64) error
+	// getMirrorPodByPod calls pod manager's GetMirrorPodByPod
+	getMirrorPodByPod func(pod *v1.Pod) (*v1.Pod, bool)
+}
+
+// NewPodKiller returns a functional PodKiller
+func NewPodKiller(kl *Kubelet) PodKiller {
+	podKiller := &podKillerWithSlice{
+		podsToKill:        make([]*kubecontainer.PodPair, 0),
```

Comment:
adding `podsToKill` plus `cond` has made this code more complex over the channel approach.  please return the the channel method.

Reply:
I have restored to yesterday's formation.